bundles/org.eclipse.jface/src/org/eclipse/jface/resource/URLImageDescriptor.java

```diff
@@ -215,12 +226,49 @@ private static URL getxURL(URL url, int zoom) {
 				if (url.getQuery() != null) {
 					file += '?' + url.getQuery();
 				}
-				return new URL(url.getProtocol(), url.getHost(), url.getPort(), file);
+				resultingURL = new URL(url.getProtocol(), url.getHost(), url.getPort(), file);
+			} catch (MalformedURLException e) {
+				Policy.getLog().log(new Status(IStatus.ERROR, Policy.JFACE, e.getLocalizedMessage(), e));
+			}
+		}
+
+		// Next we check if there's an alternate image that we should use
+		// TODO: I'm not exactly sure how we should be accessing the JFace preference
+		// Should a IPreferenceStore instance be retrieved?
+		// Or is that considered part of Eclipse RCP and not JFace?
+		if (
+		// !JFaceResources.getString(JFacePreferences.CUSTOM_RESOURCE_THEME).equals(JFacePreferences.CUSTOM_RESOURCE_THEME)
+		// &&
+		resultingURL.getFile().contains(PLUGIN_URL)) {
+			// The third path segment is the bundleID
+			String originalFilePath = resultingURL.getFile();
+			for (int i = 0; i < 2; i++) {
+				originalFilePath = originalFilePath.substring(originalFilePath.indexOf('/') + 1);
+			}
+			// TODO: The original bundleID should only be used in the case of plugin
+			// fragment image replacements.
+			// We need an additional (optional) preference that will hold the bundleID of
+			// the currently used theme plugin
+			String bundleID = originalFilePath.substring(0, originalFilePath.indexOf('/'));
+			originalFilePath = originalFilePath.substring(originalFilePath.indexOf('/'), originalFilePath.length());
+
+			// TODO: This is a workaround and should probably be removed
+			if (originalFilePath.contains("/$nl$/")) { //$NON-NLS-1$
+
+				originalFilePath = originalFilePath.replace("/$nl$/", "/"); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			try {
+				URL themedIconURL = new URL(resultingURL.getProtocol(), resultingURL.getHost(), resultingURL.getPort(),
+						PLUGIN_URL + bundleID + THEME_REPLACEMENTS_DIR + originalFilePath);
+				String found = getFilePath(themedIconURL, false);
```

Comment:
I don't think that filepath is suitable here, this further limits the usage of this.

Reply:
My intention here was to validate that the newly constructed icon URL actually resolves, and if not, fallback to the default icon. Do you know if there is a better way to see if a URL resolves, that doesn't rely on the assumption that bundles are local?

Reply:
Ah maybe instead of `getFilePath(themedIconURL, false)` I could do something like `URLImageDescriptor.getImageData(themedIconURL)`? I'm investigating [your commit](https://git.eclipse.org/r/c/platform/eclipse.platform.ui/+/170856) for help :)

Reply:
The only way that works for all cases is try open the URL, and maybe fall back to either the error or the original URL.